pkg/lib/packagekit.es6

```diff
@@ -264,3 +264,83 @@ export function watchRedHatSubscription(callback) {
             }
         );
     }
+
+/* Support for installing optional packages.
+ *
+ * First call check_optional_packages to determine whether something
+ * needs to be installed, then call install_optional_packages to
+ * actually install them.
+ *
+ * check_optional_packages resolves to either null when no packages
+ * need to be installed, or an object that can be passed to
+ * install_optional_packages.
+ *
+ * In the future, that object might contain more information about
+ * how many packages need to be installed, and how much space that
+ * will take, but, PackageKit...
+ */
+
+export function check_optional_packages(names, progress_cb) {
+    if (names.length == 0)
+        return cockpit.resolve(null);
+
+    function refresh() {
+        return cancellableTransaction("RefreshCache", [ false ], progress_cb);
+    }
+
+    function resolve() {
+        var ids = [ ];
+        return cancellableTransaction("Resolve",
+                                      [ Enum.FILTER_ARCH |
+                                        Enum.FILTER_NOT_SOURCE |
+                                        Enum.FILTER_NEWEST,
+                                        names
+                                      ],
+                                      progress_cb,
+                                      {
+                                          Package: (info, package_id) => {
+                                              console.log("P", info, package_id);
+                                              var repos = package_id.split(";")[3].split(":");
+                                              if (repos.indexOf("installed") == -1)
+                                                  ids.push(package_id);
+                                          },
+                                      }).then(() => {
+                                          if (ids.length == 0)
+                                              return null;
+                                          else
+                                              return { pkg_ids: ids };
```

Comment:
Do you rely on the special `null` case in particular? It seems more consistent to always just return `ids` even if empty. `install_packages()` below in particular deals fine with this, and it also makes it easier to log/print/check the value without always having to consider the `null` special case.

Reply:
I see the point, but with the current API, `pkg_ids` is internal and the client shouldn't access it.

Reply:
Fixed.